src/Core/Grid/Query/CategoryQueryBuilder.php

```diff
@@ -0,0 +1,203 @@
+     * @param int $contextShopId
+     * @param MultistoreContextCheckerInterface $multistoreContextChecker
+     * @param FeatureInterface $multistoreFeature
+     */
+    public function __construct(
+        Connection $connection,
+        $dbPrefix,
+        DoctrineSearchCriteriaApplicator $searchCriteriaApplicator,
+        $contextLangId,
+        $contextShopId,
+        MultistoreContextCheckerInterface $multistoreContextChecker,
+        FeatureInterface $multistoreFeature
+    ) {
+        parent::__construct($connection, $dbPrefix);
+
+        $this->contextLangId = $contextLangId;
+        $this->contextShopId = $contextShopId;
+        $this->searchCriteriaApplicator = $searchCriteriaApplicator;
+        $this->multistoreContextChecker = $multistoreContextChecker;
+        $this->multistoreFeature = $multistoreFeature;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getSearchQueryBuilder(SearchCriteriaInterface $searchCriteria)
+    {
+        $qb = $this->getQueryBuilder($searchCriteria->getFilters());
+        $qb->select('c.id_category, c.id_parent, c.active, cl.name, cl.description, cs.position');
+
+        $this->searchCriteriaApplicator
+            ->applyPagination($searchCriteria, $qb)
+            ->applySorting($searchCriteria, $qb)
+        ;
+
+        return $qb;
+    }
+
+    /**
+     * {@inheritdoc}
+     */
+    public function getCountQueryBuilder(SearchCriteriaInterface $searchCriteria)
+    {
+        $qb = $this->getQueryBuilder($searchCriteria->getFilters());
+        $qb->select('COUNT(c.id_category)');
+
+        return $qb;
+    }
+
+    /**
+     * Get generic query builder.
+     *
+     * @param array $filters
+     *
+     * @return QueryBuilder
+     */
+    private function getQueryBuilder(array $filters)
+    {
+        $qb = $this->connection
+            ->createQueryBuilder()
+            ->from($this->dbPrefix . 'category', 'c')
+            ->setParameter('context_lang_id', $this->contextLangId)
+            ->setParameter('context_shop_id', $this->contextShopId)
+        ;
+
+        $qb->leftJoin(
+            'c',
+            $this->dbPrefix . 'category_lang',
+            'cl',
+            $this->multistoreFeature->isUsed() && $this->multistoreContextChecker->isSingleShopContext() ?
+                'c.id_category = cl.id_category AND cl.id_lang = :context_lang_id AND cl.id_shop = :context_shop_id' :
+                'c.id_category = cl.id_category AND cl.id_lang = :context_lang_id AND cl.id_shop = c.id_shop_default'
+        );
+
+        $qb->leftJoin(
+            'c',
+            $this->dbPrefix . 'category_shop',
+            'cs',
+            $this->multistoreContextChecker->isSingleShopContext() ?
+                'c.id_category = cs.id_category AND cs.id_shop = :context_shop_id' :
+                'c.id_category = cs.id_category AND cs.id_shop = c.id_shop_default'
+        );
+
+        foreach ($filters as $filterName => $filterValue) {
```

Comment:
You maybe should be able to refactor this huge foreach?

Reply:
i dont think if it's really necessary, as there are plans to introducing query applications so later it probably will be refactored. i think @mickaelandrieu had some ideas about it?